src/main/java/apoc/hashing/Fingerprinting.java

```diff
@@ -0,0 +1,198 @@
+import java.util.stream.StreamSupport;
+
+public class Fingerprinting {
+
+    public static final String DIGEST_ALGORITHM = "MD5";
+
+    @Context
+    public GraphDatabaseService db;
+
+    @UserFunction
+    @Description("calculate a checksum (md5) over a node or a relationship. This deals gracefully with array properties. Two identical entities do share the same hash.")
+    public String fingerprintNodeOrRel(@Name("node_relation") Object thing, @Name(value = "propertyExcludes", defaultValue = "") List<String> excludedPropertyKeys) {
+
+        if (thing instanceof Node) {
+            return fingerprintNode((Node) thing, excludedPropertyKeys);
+        } else if (thing instanceof Relationship) {
+                return fingerprintRelationship((Relationship) thing, excludedPropertyKeys);
+        } else {
+            throw new UnsupportedOperationException("cannot fingerprint a " + thing.getClass());
+        }
+
+    }
+
+    @UserFunction
+    @Description("calculate a checksum (md5) over a the full graph. Be aware that this function does use in-memomry datastructures depending on the size of your graph.")
+    public String fingerprintGraph(@Name(value = "propertyExcludes", defaultValue = "") List<String> excludedPropertyKeys) {
+
+        return withMessageDigest(messageDigest -> {
+            // step 1: load all nodes, calc their hash and map them to id
+            Map<Long, String> idToNodeHash = db.getAllNodes().stream().collect(Collectors.toMap(
+                    Node::getId,
+                    node -> fingerprintNode(node, excludedPropertyKeys),
+                    (aLong, aLong2) -> { throw new RuntimeException(); },
+                    () -> new TreeMap<>()
+            ));
+
+            // step 2: build inverse map
+            Map<String, Long> nodeHashToId = idToNodeHash.entrySet().stream().collect(Collectors.toMap(
+                    Map.Entry::getValue,
+                    Map.Entry::getKey,
+                    (o, o2) -> { throw new RuntimeException(); },
+                    () -> new TreeMap<>()
+            ));
+
+            // step 3: iterate nodes in order of their hash (we cannot rely on internal ids)
+            nodeHashToId.entrySet().stream().forEach(entry ->{
+                messageDigest.update(entry.getKey().getBytes());
+
+                Node node = db.getNodeById(entry.getValue());
+                List<EndNodeRelationshipHashTuple> endNodeRelationshipHashTuples = StreamSupport.stream(node.getRelationships(Direction.OUTGOING).spliterator(), false)
+                        .map(relationship -> {
+                            String endNodeHash = idToNodeHash.get(relationship.getEndNodeId());
+                            String relationshipHash = fingerprintRelationship(relationship, excludedPropertyKeys);
+                            return new EndNodeRelationshipHashTuple(endNodeHash, relationshipHash);
+                        }).collect(Collectors.toList());
```

Comment:
do we have to materialize the list here in between?
we can just continue with the `.sorted().forEach(..)`